### ISSUE TITLE
is_item can be costly

### ISSUE THREAD

**Issue by Gallaecio** [scrapy/itemadapter#59]
See https://github.com/scrapy/itemloaders/issues/50

While that issue can be addressed in a different way, I am thinking that we should make some work to minimize the CPU cost of this function.

I am actually starting to consider making different item types opt-in; i.e. instead of having all supported item types enabled by default, requiring users to enable them for them to work.

I am just thinking out loud, though. I am just worried about the CPU-consumption of this function going forward, seeing as it is already able to cause a 50% additional CPU usage for a process when overused.

**Comment by elacuesta** [comment]
I remember a discussion about imports inside functions being expensive. Let's see if #60 helps in that regard.

**Comment by elacuesta** [comment]
> I am actually starting to consider making different item types opt-in; i.e. instead of having all supported item types enabled by default, requiring users to enable them for them to work.

I just realized this is actually possible at the moment, by modifying the [`ItemAdapter.ADAPTER_CLASSES`](https://github.com/scrapy/itemadapter/blob/v0.4.0/itemadapter/adapter.py#L257) class attribute, e.g. removing the ones you are sure you don't need and/or adjusting the order to break early if you know the majority of your items are of a certain kind.

**Comment by Gallaecio** [comment]
> > I am actually starting to consider making different item types opt-in; i.e. instead of having all supported item types enabled by default, requiring users to enable them for them to work.
> 
> I just realized this is actually possible at the moment, by modifying the [`ItemAdapter.ADAPTER_CLASSES`](https://github.com/scrapy/itemadapter/blob/v0.4.0/itemadapter/adapter.py#L257) class attribute, e.g. removing the ones you are sure you don't need and/or adjusting the order to break early if you know the majority of your items are of a certain kind.

Indeed, they are currently opt-out. Making them opt-in would mean making `ItemAdapter.ADAPTER_CLASSES` empty or at least less crowded by default. But it could be argued that the user experience would be worse, not to mention that making such a change in a backward-compatible way could be a pain.

**Comment by elacuesta** [comment]
Released [v0.5.0](https://github.com/scrapy/itemadapter/releases/tag/v0.5.0) to get advantage of the performance improvements. Let's keep this issue open to investigate about further improvements.
